core/src/identity/ecdsa.rs

```diff
@@ -0,0 +1,161 @@
+
+//! ECDSA keys with secp256r1 curve support.
+
+use super::error::DecodingError;
+use core::fmt;
+use p256::ecdsa::{
+    signature::{Signer, Verifier},
+    Signature, SigningKey, VerifyingKey,
+};
+
+/// An ECDSA keypair.
+#[derive(Clone)]
+pub struct Keypair {
+    secret: SecretKey,
+    public: PublicKey,
+}
+
+impl Keypair {
+    /// Generate a new random ECDSA keypair.
+    pub fn generate() -> Keypair {
+        Keypair::from(SecretKey::generate())
+    }
+
+    /// Sign a message using the private key of this keypair.
+    pub fn sign(&self, msg: &[u8]) -> Vec<u8> {
+        self.secret.sign(msg)
+    }
+
+    /// Get the public key of this keypair.
+    pub fn public(&self) -> &PublicKey {
+        &self.public
+    }
+
+    /// Get the secret key of this keypair.
+    pub fn secret(&self) -> &SecretKey {
+        &self.secret
+    }
+}
+
+impl fmt::Debug for Keypair {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Keypair")
+            .field("public", &self.public())
+            .finish()
+    }
+}
+
+/// Promote an ECDSA secret key into a keypair.
+impl From<SecretKey> for Keypair {
+    fn from(secret: SecretKey) -> Keypair {
+        let public = PublicKey(VerifyingKey::from(&secret.0));
+        Keypair { secret, public }
+    }
+}
+
+/// Demote an ECDSA keypair to a secret key.
+impl From<Keypair> for SecretKey {
+    fn from(kp: Keypair) -> SecretKey {
+        kp.secret
+    }
+}
+
+/// An ECDSA secret key.
+#[derive(Clone)]
+pub struct SecretKey(SigningKey);
+
+impl SecretKey {
+    /// Generate a new random ECDSA secret key.
+    pub fn generate() -> SecretKey {
+        SecretKey(SigningKey::random(rand::thread_rng()))
+    }
+
+    /// Sign a message with this secret key, producing a DER-encoded ECDSA signature.
+    pub fn sign(&self, msg: &[u8]) -> Vec<u8> {
+        self.0.sign(msg).to_der().as_bytes().to_owned()
+    }
+}
+
+impl fmt::Debug for SecretKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "SecretKey")
+    }
+}
+
+/// An ECDSA public key.
+#[derive(Clone, PartialEq, Eq)]
+pub struct PublicKey(VerifyingKey);
+
+impl PublicKey {
+    /// Verify an ECDSA signature on a message using the public key.
+    pub fn verify(&self, msg: &[u8], sig: &[u8]) -> bool {
+        let sig = match Signature::from_der(sig) {
+            Ok(sig) => sig,
+            Err(_) => return false,
+        };
+        self.0.verify(msg, &sig).is_ok()
+    }
+
+    /// Encode a public key into a DER encoded byte buffer as defined by SEC1 standard.
+    pub fn encode_der(&self) -> Vec<u8> {
+        self.0.to_encoded_point(false).as_bytes().to_owned()
+    }
+
+    // Decode a public key from a DER encoded byte buffer as defined for SEC1 standard.
+    pub fn decode_der(k: &[u8]) -> Result<PublicKey, DecodingError> {
+        VerifyingKey::from_sec1_bytes(k)
+            .map_err(|_| DecodingError::new("failed to parse ecdsa p256 public key"))
```

Comment:
Could `?` after this to avoid the last `.map`.

Reply:
We are dropping the source error here. Is that what we do for other keypairs as well? I think we should include the source error message.

Reply:
The second `map` has been inserted to map from `Result<VerifyingKey>` into a `Result<PublicKey>`.

With PublicKey defined as : `struct PublicKey(VerifyingKey)`

Reply:
This should have been resolved.

Where possible the errors are propagated using `source()`. The is an error that cannot be stored within the source because it doesn't implement `Error`. In this case I've added some details to the error message.

```
        let enc_pt = EncodedPoint::from_bytes(k).map_err(|_| {   // <- this doesn't implement error on `no_std` builds
            DecodingError::new("failed to parse ecdsa p256 public key, bad point encoding")
        })?;
```